### PR TITLE
feat: Add `BoolTokens` mixin for strict/fgpyo-style bool parsing

### PR DESCRIPTION
**Pull request by msto**
## Summary

Closes #75

One of two features to bring `Metric`'s parsing to parity with fgpyo. 

Pydantic accepts a broad set of strings as booleans (`true/t/yes/y/on/1` and false counterparts). This PR updates `Metric`'s boolean parsing to be consistent with fgpyo by default - `{true, t, 1}`/`{false, f, 0}` are case-insensitively matched to `True`/`False`, and any other values raise a `ValidationError`.

### CODE REVIEW

File: fgmetric/converters/_bool_tokens.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    The default token set mirrors `fgpyo.util.types.parse_bool`: `{"true", "t", "1"}` for `True`
+    and `{"false", "f", "0"}` for `False`. This is stricter than Pydantic, which would also accept
+    `yes`/`y`/`on`/`no`/`n`/`off`. Rejecting those guards against a misaligned or wrong-type column
+    being silently coerced, and gives tools ported from fgpyo identical accept/reject behavior.
+
+    This mixin is read-side only: serialization is unchanged (a `bool` still serializes to
+    `"True"`/`"False"`, matching fgpyo).
+
+    Tokens are matched case-insensitively, so they may be configured in any case. `true_tokens` and
+    `false_tokens` must be disjoint; declaring a token in both is rejected at class definition.
```
Comment (msto) [inline fgmetric/converters/_bool_tokens.py]:
```suggestion
    This mixin is read-side only; a `bool` still serializes to`"True"`/`"False"`.

    Tokens are matched case-insensitively. `true_tokens` and
    `false_tokens` must be disjoint; declaring a token in both is rejected at class definition.
```

File: fgmetric/converters/_bool_tokens.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    """
+
+    # Public, user-specified token sets.
+    true_tokens: ClassVar[frozenset[str]] = frozenset({"true", "t", "1"})
+    false_tokens: ClassVar[frozenset[str]] = frozenset({"false", "f", "0"})
+
+    _bool_fieldnames: ClassVar[set[str]]
+    # Internal: lowercase copies of the user-specified token sets, used for the actual matching.
+    _true_tokens: ClassVar[frozenset[str]]
+    _false_tokens: ClassVar[frozenset[str]]
```
Comment (msto) [inline fgmetric/converters/_bool_tokens.py]:
```suggestion
    # Internal: lowercase copies of the user-specified token sets, used for the actual matching.
    _true_tokens: ClassVar[frozenset[str]]
    _false_tokens: ClassVar[frozenset[str]]
    
    _bool_fieldnames: ClassVar[set[str]]
```

File: fgmetric/converters/_bool_tokens.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        false_tokens: The strings resolved to `False`. Defaults to `frozenset({"false", "f", "0"})`.
+
+    Note:
+        `bool` is a subclass of `int`, but only fields annotated exactly as `bool` or `bool | None`
+        are affected; `int` fields are untouched.
+
+        When combined with `NullSentinels`, the model-level null substitution runs first, so an
+        empty `bool | None` cell is converted to `None` before this validator sees it (provided
+        `""` is a configured null sentinel). Without such a sentinel, an empty cell on a `bool`
+        field is rejected, since `""` is in neither token set.
```
Comment (ameynert) [inline fgmetric/converters/_bool_tokens.py:51]: [praise](https://pullpo.io/cc?l=praise) 
Appreciate this thorough documentation!

File: fgmetric/converters/_bool_tokens.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        ```
+
+        Customize the accepted tokens:
+
+        ```python
+        class MyMetric(Metric):
+            true_tokens = frozenset({"yes", "y"})
+            false_tokens = frozenset({"no", "n"})
+            flag: bool
+        ```
```
Comment (ameynert) [inline fgmetric/converters/_bool_tokens.py:74]: [suggestion](https://pullpo.io/cc?l=suggestion) 

```suggestion

        MyMetric.model_validate({"flag": "1"}) # -> ValidationError
        MyMetric.model_validate({"flag": "yes"}).flag # -> True
        ```
```

File: fgmetric/converters/_bool_tokens.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    The default token set mirrors `fgpyo.util.types.parse_bool`: `{"true", "t", "1"}` for `True`
+    and `{"false", "f", "0"}` for `False`. This is stricter than Pydantic, which would also accept
+    `yes`/`y`/`on`/`no`/`n`/`off`. Rejecting those guards against a misaligned or wrong-type column
+    being silently coerced, and gives tools ported from fgpyo identical accept/reject behavior.
+
+    This mixin is read-side only; a `bool` still serializes to `"True"`/`"False"`.
+
+    Tokens are matched case-insensitively. `true_tokens` and `false_tokens` must be disjoint;
+    declaring a token in both is rejected at class definition.
```
Comment (ameynert) [inline fgmetric/converters/_bool_tokens.py:33]: [suggestion](https://pullpo.io/cc?l=suggestion) 

```suggestion
    declaring a token in both is rejected at class definition.
    
    Non-string inputs are left to pydantic.
```

**Review by ameynert — COMMENTED** [review]
`list[bool] / list[bool | None]` element values silently bypass the strict token set. Confirmed empirically against the branch:

```python
  class M(Metric):
      flags: list[bool]
      flag: bool

  M.model_validate({"flag": "yes"})                 # ValidationError  (strict, correct)
  M.model_validate({"flags": "yes,on,no"}).flags    # -> [True, True, False]  ← broad coercion!
```

  `is_bool(list[bool])` is `False`, so list-of-bool fields never enter `_bool_fieldnames`; DelimitedList._split_lists splits the cell and pydantic's default broad coercion validates each element — accepting exactly the yes/on/off tokens this PR exists to reject. So a scalar `bool` column and a `list[bool]` column on the same `Metric` disagree on what's a valid boolean.
  
Options:
  1. Extend token parsing to bool-typed list elements (apply tokenization per element when `has_origin(annotation, list)` and the element type is `bool`), or
  2. Explicitly document the limitation in the `BoolTokens` docstring and add a test pinning the current `list[bool]` behavior so it's a deliberate, visible choice
  rather than an accident.

**Review by ameynert — CHANGES_REQUESTED** [review]

File: fgmetric/converters/_bool_tokens.py [hunk 6]
```diff
@@ -31,6 +32,11 @@ class BoolTokens(BaseModel):
     Tokens are matched case-insensitively. `true_tokens` and `false_tokens` must be disjoint;
     declaring a token in both is rejected at class definition.
 
+    Fields whose elements are bool - `list[bool]` and `list[bool | None]` - are tokenized per
+    element, so a list-of-bool column and a scalar `bool` column on the same model agree on what
+    counts as a valid boolean. Non-string values (and non-string list elements, such as a `None`
+    from an empty `list[bool | None]` cell) are left to Pydantic.
```
Comment (ameynert) [inline fgmetric/converters/_bool_tokens.py:38]: [question](https://pullpo.io/cc?l=question) 
How would something weird `list[bool | str]` be treated? Is that even allowed?

File: fgmetric/converters/_bool_tokens.py [hunk 7]
```diff
@@ -108,17 +127,28 @@ def _require_disjoint_tokens(cls) -> None:
     @field_validator("*", mode="before")
     @classmethod
     def _parse_bool_tokens(cls, value: Any, info: ValidationInfo) -> Any:
-        """Resolve string values on bool fields to `True`/`False` from the configured tokens."""
+        """Resolve string values on bool (and bool-list) fields from the configured tokens."""
```
Comment (ameynert) [inline fgmetric/converters/_bool_tokens.py:130]: [suggestion](https://pullpo.io/cc?l=suggestion) 

Claude spotted this - worth a comment to ensure future refactors keep `_split_list` running before this.

```
        # On `Metric`, `DelimitedList` runs first and has already split a bool-list cell into a
        # list of string elements by the time this validator sees it; resolve each element. Real
        # `None` elements (e.g. an empty cell in a `list[bool | None]`) are left for Pydantic.
        #
        # NB: This correctness hinges on `DelimitedList`'s before-validator running ahead of this
        # one (it must split the cell into a list before we can tokenize elements). If that order
        # ever flips, a `list[bool]` cell reaches Pydantic as an unsplit string and falls back to
        # broad coercion (silently accepting `yes`/`on`/...). `test_metric_bool_list_is_strict`
        # is the regression guard.
```

File: tests/test_typing_extensions.py [hunk 8]
```diff
@@ -107,6 +108,39 @@ def test_is_not_bool(annotation: TypeAnnotation) -> None:
...
+)
+def test_is_bool_list(annotation: TypeAnnotation) -> None:
+    """Should identify lists of `bool` / `bool | None`, even within an Optional."""
+    assert is_bool_list(annotation)
+
+
+@pytest.mark.parametrize(
+    "annotation",
+    [
+        bool,  # a scalar bool is not a list
```
Comment (ameynert) [inline tests/test_typing_extensions.py:131]: [suggestion](https://pullpo.io/cc?l=suggestion) 
Add a test where the input `TypeAnnotation` is `None` (same with `test_is_not_bool` above) - that should return `False` for both.

File: fgmetric/metric.py
Comment (ameynert) [inline fgmetric/metric.py:56]: [suggestion](https://pullpo.io/cc?l=suggestion) 
```suggestion
        true_tokens: The set of input strings parsed as `True` on `bool` fields and `list[bool]`
            elements during validation. Defaults to `frozenset({"true", "t", "1"})`.
        false_tokens: The set of input strings parsed as `False` on `bool` fields and `list[bool]`
            elements during validation. Defaults to `frozenset({"false", "f", "0"})`.
```

**Review by ameynert — APPROVED** [review]
